### PR TITLE
Assignment3

### PR DESCRIPTION
**Pull request by dhoxie**
@xiokd

### CODE REVIEW

**Comment by dhoxie** [comment]
## Assignment

For this assignment do the following:

- Create *two* interfaces. One to represent the `JokeService` and one interface to represent outputting the joke to the screen (`Console.WriteLine`). ✔
- The `JokeService` will need to have the interface applied to it. ✔
- Create an implementation of the output interface that writes out the joke to the console. ✔
- Implement the `Jester` class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
- The `Jester` class `TellJoke()` method should retrieve a joke from the `JokeService`. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
- Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ❌ (Currently at ~86%)
- Enable nullability in the .csproj files and ensure no errors or warnings: ✔

## Fundamentals
- Ensure you enable:
  - nullable reference types is enabled ✔
  - net6 targetted  ✔
  - C# 10.0  ✔
  - and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

## Notes
- Unit test coverage for the Jester class is currently at 86%.  We plan to improve this and update the PR.
- Did not attempt extra credit

**Comment by dhoxie** [comment]
## Assignment - Updated Review

For this assignment do the following:

- Create *two* interfaces. One to represent the `JokeService` and one interface to represent outputting the joke to the screen (`Console.WriteLine`). ✔
- The `JokeService` will need to have the interface applied to it. ✔
- Create an implementation of the output interface that writes out the joke to the console. ✔
- Implement the `Jester` class. It should take in both interfaces as dependencies. These dependencies should be null checked. ✔
- The `Jester` class `TellJoke()` method should retrieve a joke from the `JokeService`. If the joke contains "Chuck Norris", skip it and get another. The joke should be written to the output dependency. ✔
- Unit test the Jester class. Code coverage should be above 90% for this class. Moq will make this significantly easier. ✔
- Enable nullability in the .csproj files and ensure no errors or warnings: ✔

## Fundamentals
- Ensure you enable:
  - nullable reference types is enabled ✔
  - net6 targetted  ✔
  - C# 10.0  ✔
  - and enabled .NET analyzers for both projects ✔
- For this assignment, always use `Assert.AreEqual<T>()` (the generic version)  ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

## Notes
- Did not attempt extra credit

File: CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj [hunk 1]
```diff
@@ -2,8 +2,10 @@
 
   <PropertyGroup>
     <TargetFramework>net6.0</TargetFramework>
-
+    <LangVersion>10.0</LangVersion>
+    <Nullable>enable</Nullable>
     <IsPackable>false</IsPackable>
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny.Tests/CanHazFunny.Tests.csproj:7]:
```suggestion
    <IsPackable>false</IsPackable>
    <EnableNETAnalyzers>true</EnableNETAnalyzers>
```

File: CanHazFunny/CanHazFunny/PrintToConsoleService.cs [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+namespace CanHazFunny
+{
+    public class PrintToConsoleService : IPrint
+    {
+
+        public void Print(string text)
+        {
+            if (text == null) throw new ArgumentNullException(nameof(text));
+            if (text.Length == 0) throw new ArgumentException(nameof(text));
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/PrintToConsoleService.cs:11]:
```suggestion
            if(string.IsNullOrWhiteSpace(text)) throw new ArgumentException($"{nameof(text)} should not be null or empty");
```
You can use `IsNullOrWhiteSpace` or `IsNullOrEmpty` to do this in one line. It is also a good idea to give a better description with your `ArgumentException`.

File: CanHazFunny/CanHazFunny/Jester.cs [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        {
+            _JokeService = jokeService ?? throw new ArgumentNullException(nameof(jokeService));
+            _PrintService = printService ?? throw new ArgumentNullException(nameof(printService));
+        }
+
+        public void TellJoke()
+        {
+            string joke = _JokeService.GetJoke();
+
+            if (joke == null) { throw new ArgumentNullException(nameof(joke)); }
```
Comment (mmwoodfo) [inline CanHazFunny/CanHazFunny/Jester.cs:20]:
```suggestion
            if (joke == null) { throw new ArgumentNullException(nameof(joke)); }
```
You shouldn't need this here

File: Lecture/Lecture.Tests/StudentTests.cs [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+        [TestMethod]
+        public void Student_AssignInteger42_Success()
+        {
+            Student student = new Student("Princess Buttercup");
+            student.Grade = (Grade)42;
+
+            Assert.AreEqual<Grade>(student.Grade, Grade.A);
+        }
+    }
+}
```
Comment (mmwoodfo) [inline Lecture/Lecture.Tests/StudentTests.cs:30]: Try to make sure you only include files related to the assignment

**Pull request closed without merging**